nexus/src/db/pagination.rs

```diff
@@ -61,3 +63,94 @@ where
+pub fn paginated_multicolumn<T, C1, C2, M1, M2>(
+    table: T,
+    (c1, c2): (C1, C2),
+    pagparams: &DataPageParams<'_, (M1, M2)>,
+) -> BoxedQuery<T>
+where
+    // T is a table which can create a BoxedQuery.
+    T: diesel::Table,
+    T: query_methods::BoxedDsl<'static, Pg, Output = BoxedQuery<T>>,
+    // C1 & C2 are columns which appear in T.
+    C1: 'static + Column + Copy + ExpressionMethods + AppearsOnTable<T>,
+    C2: 'static + Column + Copy + ExpressionMethods + AppearsOnTable<T>,
+    // Required to compare the columns with the marker types.
+    C1::SqlType: SqlType,
+    C2::SqlType: SqlType,
+    M1: Clone + AsExpression<C1::SqlType>,
+    M2: Clone + AsExpression<C2::SqlType>,
+    // Necessary for "query.order(c1.desc())"
+    BoxedQuery<T>: query_methods::OrderDsl<Desc<C1>, Output = BoxedQuery<T>>,
+    // Necessary for "query.order(...).then_order_by(c2.desc())"
+    BoxedQuery<T>:
+        query_methods::ThenOrderDsl<Desc<C2>, Output = BoxedQuery<T>>,
+    // Necessary for "query.order(c1.asc())"
+    BoxedQuery<T>: query_methods::OrderDsl<Asc<C1>, Output = BoxedQuery<T>>,
+    // Necessary for "query.order(...).then_order_by(c1.asc())"
+    BoxedQuery<T>: query_methods::ThenOrderDsl<Asc<C2>, Output = BoxedQuery<T>>,
+
+    // We'd like to be able to call:
+    //
+    //  c1.eq(v1).and(c2.gt(v2))
+    //
+    // This means "c1.eq(v1)" must implement BoolExpressionMethods, and
+    // satisfy the requirements of the ".and" method.
+    //
+    // The LHS (c1.eq(v1)) must be a boolean expression:
+    Eq<C1, M1>: Expression<SqlType = Bool>,
+    // The RHS (c2.gt(v2)) must be a boolean expression:
+    Gt<C2, M2>: Expression<SqlType = Bool>,
+    // Putting it together, we should be able to filter by LHS.and(RHS):
+    BoxedQuery<T>: query_methods::FilterDsl<
+        And<Eq<C1, M1>, Gt<C2, M2>>,
+        Output = BoxedQuery<T>,
+    >,
+
+    // We'd also like to be able to call:
+    //
+    //  c1.eq(v1).and(c2.lt(v2))
+    //
+    // We've already defined the bound on the LHS, so we add the equivalent
+    // bounds on the RHS for the "Less than" variant.
+    Lt<C2, M2>: Expression<SqlType = Bool>,
+    BoxedQuery<T>: query_methods::FilterDsl<
+        And<Eq<C1, M1>, Lt<C2, M2>>,
+        Output = BoxedQuery<T>,
+    >,
+
+    // Necessary for "query.or_filter(c1.gt(v1))"
+    BoxedQuery<T>:
+        query_methods::OrFilterDsl<Gt<C1, M1>, Output = BoxedQuery<T>>,
+    // Necessary for "query.or_filter(c1.lt(v1))"
+    BoxedQuery<T>:
+        query_methods::OrFilterDsl<Lt<C1, M1>, Output = BoxedQuery<T>>,
+{
+    let mut query = table.into_boxed().limit(pagparams.limit.get().into());
+    let marker = pagparams.marker.map(|m| m.clone());
+    match pagparams.direction {
+        dropshot::PaginationOrder::Ascending => {
+            if let Some((v1, v2)) = marker {
+                query = query.filter(c1.eq(v1.clone()).and(c2.gt(v2)));
+                query = query.or_filter(c1.gt(v1));
```

Comment:
Two functions provided by Diesel actually made these bounds a little easier for me to write, but admittedly different from your original implementation: 

https://docs.diesel.rs/master/diesel/prelude/trait.QueryDsl.html#method.or_filter
https://docs.diesel.rs/master/diesel/prelude/trait.QueryDsl.html#method.then_order_by

These are basically "optional versions" of filter and order, respectively, that can reduce the size of arguments passed.

They should be semantically equivalent to the "single-call, larger argument" version, but the trait bounds get a little less hairy.